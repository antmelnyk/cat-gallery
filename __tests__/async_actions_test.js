import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import nock from 'nock'
import fetch from 'isomorphic-fetch'

import {
  fetchCategories,
  selectCategory,
  fetchImages
} from '../src/store/actions'

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

describe('Async actions', () => {
  let store;
  const fetchCategoriesData = [
    {
      'id': 1,
      'name': 'Test'
    }
  ];
  const fetchImagesData = [];

  beforeEach(() => {
    store = mockStore({
      selectedCategory: 1,
      pagesLoaded: 0
    });
  })

  afterEach(() => {
    nock.cleanAll();
  })

  it('dispatches FETCH_TODO_SUCCESS when categories are fetched', () => {
    nock('https://api.thecatapi.com')
      .get('/v1/categories')
      .reply(200, fetchCategoriesData);

    return store.dispatch(fetchCategories())
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      })
  })

  it('dispatches FETCH_TODO_FAILURE when categories are not fetched', () => {
    nock('https://api.thecatapi.com')
      .get('/v1/categories')
      .reply(404);

    return store.dispatch(fetchCategories())
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      })
  })

  it('dispatches FETCH_IMAGES_SUCCESS when category is selected', () => { 
    nock('https://api.thecatapi.com')
      .get('/v1/images/search?order=ASC&category_ids=1&page=0&limit=8')
      .reply(200, fetchImagesData);
      
    return store.dispatch(selectCategory(1))
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      })
  })

  it('dispatches FETCH_IMAGES_SUCCESS when images are fetched', () => {
    nock('https://api.thecatapi.com')
      .get('/v1/images/search?order=ASC&category_ids=1&page=0&limit=8')
      .reply(200, fetchImagesData);

    return store.dispatch(fetchImages())
      .then(() => {
        expect(store.getActions()).toMatchSnapshot();
      })
  })
})
