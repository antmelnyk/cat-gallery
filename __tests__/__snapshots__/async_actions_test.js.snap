// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Async actions dispatches FETCH_IMAGES_SUCCESS when category is selected 1`] = `
Array [
  Object {
    "categoryId": 1,
    "type": "SET_SELECTED_CATEGORY",
  },
  Object {
    "type": "FETCH_IMAGES_REQUEST",
  },
  Object {
    "images": Array [],
    "type": "FETCH_IMAGES_SUCCESS",
  },
]
`;

exports[`Async actions dispatches FETCH_IMAGES_SUCCESS when images are fetched 1`] = `
Array [
  Object {
    "type": "FETCH_IMAGES_REQUEST",
  },
  Object {
    "images": Array [],
    "type": "FETCH_IMAGES_SUCCESS",
  },
]
`;

exports[`Async actions dispatches FETCH_TODO_FAILURE when categories are not fetched 1`] = `
Array [
  Object {
    "type": "FETCH_CATEGORIES_REQUEST",
  },
  Object {
    "error": [Error: Something went wrong. Please, try again later.],
    "type": "FETCH_REQUEST_FAILURE",
  },
]
`;

exports[`Async actions dispatches FETCH_TODO_SUCCESS when categories are fetched 1`] = `
Array [
  Object {
    "type": "FETCH_CATEGORIES_REQUEST",
  },
  Object {
    "categories": Array [
      Object {
        "id": 1,
        "name": "Test",
      },
    ],
    "type": "FETCH_CATEGORIES_SUCCESS",
  },
]
`;
